File: ios/ConverseNotificationExtension/Spam.swift
//
//  Spam.swift
//  ConverseNotificationExtension
//
//  Created by Alex Risch on 6/25/24.
//

import Foundation
import XMTP
import Alamofire

func getSenderSpamScore(address: String, apiURI: String?) async -> Double {
  var senderSpamScore: Double = 0.0
  if (apiURI != nil && !apiURI!.isEmpty) {
    let senderSpamScoreURI = "\(apiURI ?? "")/api/spam/senders/batch"
    do {
      let response = try await withUnsafeThrowingContinuation { continuation in
        AF.request(senderSpamScoreURI, method: .post, parameters: ["sendersAddresses": [address]], encoding: JSONEncoding.default, headers: nil).validate().responseData { response in
          if let data = response.data {
            continuation.resume(returning: data)
            return
          }
          if let err = response.error {
            continuation.resume(throwing: err)
            return
          }
          fatalError("should not get here")
        }
      }
      
      if let json = try JSONSerialization.jsonObject(with: response, options: []) as? [String: Any] {
        if let score = json[address] as? Double {
          senderSpamScore = score
        }
      }
    } catch {
      print(error)
    }
  }
  return senderSpamScore
}

func computeSpamScoreConversation(address: String, message: String?, contentType: String, apiURI: String?) async -> Double {
  let spamScore: Double = await getSenderSpamScore(address: address, apiURI: apiURI)
  let messageSpamScore = getMessageSpamScore(message: message, contentType: contentType)
  return spamScore + messageSpamScore
}

func getMessageSpamScore(message: String?, contentType: String) -> Double {
  var spamScore: Double = 0
  if contentType.starts(with: "xmtp.org/text:") {
    if let unwrappedMessage = message {
      if containsURL(input: unwrappedMessage) {
        spamScore += 1
      }
      
      if containsRestrictedWords(in: unwrappedMessage) {
        spamScore += 1
      }
    }
  }
  return spamScore
}

let restrictedWords = [
  "Coinbase",
  "Airdrop",
  "voucher",
  "offer",
  "restriction",
  "Congrats",
  
  "$SHIB",
  "$AERO"
]

func containsRestrictedWords(in searchString: String) -> Bool {
  // Convert the search string to lowercase
  let lowercasedSearchString = searchString.lowercased()
  
  // Check each restricted word in the lowercase search string
  for word in restrictedWords {
    if lowercasedSearchString.contains(word.lowercased()) {
      return true
    }
  }
  return false
}

func computeSpamScoreV3Welcome(client: XMTP.Client, conversation: XMTP.Conversation, apiURI: String?) async -> Double {
  do {
//    try await client.preferences.syncConsent()
//    // Probably an unlikely case until consent proofs for groups exist
//    do {
//      let convoState = try await client.preferences.conversationState(conversationId: conversation.id)
//      let convoAllowed = convoState == .allowed
//      if convoAllowed {
//        return -1
//      }
//    } catch {
//      
//    }

    if case .group(let group) = conversation {
      let inviterInboxId = try group.addedByInboxId()
      let inviterState = try await client.preferences.inboxIdState(inboxId: inviterInboxId)

      let inviterAllowed = inviterState == .allowed
      if inviterAllowed {
        return -1
      }
      let inviterDenied = inviterState == .denied
      if inviterDenied {
        return 1
      }
      let members = try await conversation.members()
      var anyDenied = false
      var anyAllowed = false
      if let inviterAddresses = members.first(where: {$0.inboxId == inviterInboxId})?.addresses {

        for address in inviterAddresses {
          let addressState = try await client.preferences.addressState(address: address)
          if addressState == .denied {
            anyDenied = true
          }
          if addressState == .allowed {
            anyAllowed = true
          }
        }
        if anyDenied {
          return 1
        }

        if anyAllowed {
          return -1
        }

        if let firstAddress = inviterAddresses.first {
          let senderSpamScore = await getSenderSpamScore(address: firstAddress, apiURI: apiURI)
            return senderSpamScore
        }
      }
    } else if case .dm(let dm) = conversation {
        print("It's a DM with details: \(dm)")
      
    }
  } catch {
    sentryTrackError(error: error, extras: ["message": "Failed to compute Spam Score for V3 Welcome"])
    return 0
  }

  return 0
}

func computeSpamScoreV3Message(client: XMTP.Client, conversation: XMTP.Conversation, decodedMessage: DecodedMessage, apiURI: String?) async -> Double {
  do {
    
//    try await client.preferences.syncConsent()
    let groupDenied = try await client.preferences.conversationState(conversationId: conversation.id) == .denied
    if groupDenied {
      // Network consent will override other checks
      return 1
    }
    let senderInboxId = decodedMessage.senderInboxId
    let senderDenied = try await client.preferences.inboxIdState(inboxId: senderInboxId) == .denied
    if senderDenied {
      // Network consent will override other checks
      return 1
    }
    
    let senderAllowed = try await client.preferences.inboxIdState(inboxId: senderInboxId) == .allowed
    if senderAllowed {
      // Network consent will override other checks
      return -1
    }
    
    
    let convoAllowed = try await client.preferences.conversationState(conversationId: conversation.id) == .allowed
    if convoAllowed {
      // Network consent will override other checks
      return -1
    }
    
    if case .group(let group) = conversation {

      if let senderAddresses = try await group.members.first(where: {$0.inboxId == senderInboxId})?.addresses {
        for address in senderAddresses {
          if try await client.preferences.addressState(address: address) == .denied {
            return 1
          }
        }
        for address in senderAddresses {
          if try await client.preferences.addressState(address: address) == .allowed {
            return -1
          }
        }
      }
    } else if case .dm(let dm) = conversation {
      let peer = try dm.peerInboxId
      
      if try await client.preferences.inboxIdState(inboxId: peer) == .allowed {
        return -1
      }
      if try await client.preferences.inboxIdState(inboxId: peer) == .denied {
        return 1
      }
    }
  } catch {
    //
    sentryTrackError(error: error, extras: ["message": "Failed to compute Spam Score for V3 Message"])
  }
  let contentType = getContentTypeString(type: decodedMessage.encodedContent.type)
  
  let messageContent = String(data: decodedMessage.encodedContent.content, encoding: .utf8)
  let messageSpamScore = getMessageSpamScore(message: messageContent, contentType: contentType)
  
  return messageSpamScore
}

File: android/app/src/main/java/com/converse/Spam.kt
import android.content.Context
import android.util.Log
import com.android.volley.Request
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.converse.containsURL
import com.converse.getAsyncStorage
import com.converse.getMmkv
import com.converse.xmtp.getContentTypeString
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.suspendCancellableCoroutine
import org.json.JSONObject
import org.xmtp.android.library.Client
import org.xmtp.android.library.DecodedMessage
import org.xmtp.android.library.Group
import java.util.HashMap
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import org.web3j.crypto.Keys
import org.xmtp.android.library.Dm

val restrictedWords = listOf(
    "Coinbase",
    "Airdrop",
    "voucher",
    "offer",
    "restriction",
    "Congrats",
    "\$SHIB",
    "\$AERO"
)

fun containsRestrictedWords(searchString: String): Boolean {
    // Convert the search string to lowercase
    val lowercasedSearchString = searchString.lowercase()

    // Check each restricted word in the lowercase search string
    for (word in restrictedWords) {
        if (lowercasedSearchString.contains(word.lowercase())) {
            return true
        }
    }
    return false
}

fun getMessageSpamScore(message: String?, contentType: String): Int {
    var spamScore = 0
    if (contentType.startsWith("xmtp.org/text:") && message?.let { containsURL(it) } == true) {
        spamScore += 1
    }
    if (message !== null && containsRestrictedWords(message)) {
        spamScore += 1
    }
    return spamScore
}

fun computeDmSpamScore(address: String, message: String?, contentType: String, apiURI: String?, appContext: Context): Double {
    var senderScore = runBlocking {
        getSenderSpamScore(appContext, address, apiURI);
    }
    return senderScore + getMessageSpamScore(message, contentType)
}

suspend fun computeSpamScoreGroupMessage(client: Client, group: Group, decodedMessage: DecodedMessage, apiURI: String?): Int {
    val senderSpamScore = 0
    try {
        client.syncConsent()
        val consentList = client.preferences.consentList
        val groupBlocked = isConversationIdBlocked(group.id, consentList)
        if (groupBlocked) {
            // Network consent will override other checks
            return 1
        }

        val senderInboxId = decodedMessage.senderAddress
        val senderBlocked = isInboxIdBlocked(senderInboxId, consentList)
        if (senderBlocked) {
            // Network consent will override other checks
            return 1
        }

        val senderAllowed = isInboxIdAllowed(senderInboxId, consentList)
        if (senderAllowed) {
            // Network consent will override other checks
            return -1
        }

        val groupAllowed = isConversationIdAllowed(group.id, consentList)
        if (groupAllowed) {
            // Network consent will override other checks
            return -1
        }

        val senderAddresses = group.members().find { it.inboxId == senderInboxId }?.addresses
        if (senderAddresses != null) {
            for (address in senderAddresses) {
                if (isAddressBlocked(Keys.toChecksumAddress(address), consentList)) {
                    return 1
                }
            }
            for (address in senderAddresses) {
                if (isAddressAllowed(Keys.toChecksumAddress(address), consentList)) {
                    return -1
                }
            }
        }

        // TODO: Handling for inbox Id
        // spamScore = getSenderSpamScore(address: senderInboxId, apiURI: apiURI)

    } catch (e: Exception) {
    }

    val contentType = getContentTypeString(decodedMessage.encodedContent.type)

    var messageContent: String? = null
    if (contentType.startsWith("xmtp.org/text:")) {
        messageContent = decodedMessage.encodedContent.content.toStringUtf8()
    }

    val messageSpamScore = getMessageSpamScore(messageContent, contentType)

    return senderSpamScore + messageSpamScore
}

suspend fun computeSpamScoreDmWelcome(appContext: Context, client: Client, dm: Dm, apiURI: String?): Double {
    try {
        val consentList = client.preferences.consentList
        // Probably an unlikely case until consent proofs for groups exist
        val groupAllowed = isConversationIdAllowed(dm.id, consentList)
        if (groupAllowed) {
            return -1.0
        }

        val peerInboxId = dm.peerInboxId
        val peerAllowed = isInboxIdAllowed(peerInboxId, consentList)
        if (peerAllowed) {
            return -1.0
        }

        val peerDenied = isInboxIdBlocked(peerInboxId, consentList)
        if (peerDenied) {
            return 1.0
        }
        val members = dm.members()
        for (member in members) {
            if (member.inboxId == peerInboxId) {
                val firstAddress = member.addresses.first()
                val senderSpamScore = getSenderSpamScore(
                    appContext = appContext,
                    address = Keys.toChecksumAddress(firstAddress),
                    apiURI = apiURI
                )
                return senderSpamScore
            }
        }
        return 0.0
    } catch (e: Exception) {
        return 0.0
    }
}


suspend fun computeSpamScoreGroupWelcome(appContext: Context, client: Client, group: Group, apiURI: String?): Double {
    try {
        val consentList = client.preferences.consentList
        // Probably an unlikely case until consent proofs for groups exist
        val groupAllowed = isConversationIdAllowed(group.id, consentList)
        if (groupAllowed) {
            return -1.0
        }

        val inviterInboxId = group.addedByInboxId()
        val inviterAllowed = isInboxIdAllowed(inviterInboxId, consentList)
        if (inviterAllowed) {
            return -1.0
        }

        val inviterDenied = isInboxIdBlocked(inviterInboxId, consentList)
        if (inviterDenied) {
            return 1.0
        }
        val members = group.members()

        for (member in members) {
            if (member.inboxId == inviterInboxId) {
                member.addresses?.forEach { address ->
                    val ethereumAddress = Keys.toChecksumAddress(address)
                    if (isAddressBlocked(ethereumAddress, consentList)) {
                        return 1.0
                    }
                }

                member.addresses?.forEach { address ->
                    val ethereumAddress = Keys.toChecksumAddress(address)
                    if (isAddressAllowed(ethereumAddress, consentList)) {
                        return -1.0
                    }
                }

                member.addresses?.firstOrNull()?.let { firstAddress ->
                    val senderSpamScore = getSenderSpamScore(
                        appContext = appContext,
                        address = Keys.toChecksumAddress(firstAddress),
                        apiURI = apiURI
                    )
                    return senderSpamScore
                }
            }
        }
    } catch (e: Exception) {
        return 0.0
    }
    return 0.0
}

suspend fun getSenderSpamScore(appContext: Context, address: String, apiURI: String?): Double {
    val senderSpamScoreURI = "$apiURI/api/spam/senders/batch"
    val params: MutableMap<String?, Any> = HashMap()
    params["sendersAddresses"] = arrayOf(address);

    val parameters = JSONObject(params as Map<*, *>?)

    return suspendCancellableCoroutine { continuation ->
        val jsonRequest = JsonObjectRequest(Request.Method.POST, senderSpamScoreURI, parameters, {
            var result = 0.0;
            if (it.has(address)){
                result = (it[address] as Int).toDouble()
            }
            continuation.resume(result)

        }) { error ->
            error.printStackTrace()
            continuation.resumeWithException(error)
        }

        Volley.newRequestQueue(appContext).add(jsonRequest)
    }

}